storage-proofs/core/src/gadgets/insertion.rs

```diff
@@ -69,6 +86,172 @@ pub fn insert<E: Engine, CS: ConstraintSystem<E>>(
     Ok(result)
 }
 
+pub fn insert_4<E: Engine, CS: ConstraintSystem<E>>(
+    cs: &mut CS,
+    element: &AllocatedNum<E>,
+    bits: &[Boolean],
+    elements: &[AllocatedNum<E>],
+) -> Result<Vec<AllocatedNum<E>>, SynthesisError> {
+    assert_eq!(elements.len() + 1, 4);
+    assert_eq!(bits.len(), 2);
+
+    /*
+    To insert A into [b, c, d] at position n of bits, represented by booleans [b0, b1, b2].
+    n [b0, b1] pos 0 1 2 3
+    0 [0, 0]       A b c d
+    1 [1, 0]       b A c d
+    2 [0, 1]       b c A d
+    3 [1, 1]       b c d A
+
+    A = element
+    b = elements[0]
+    c = elements[1]
+    d = elements[2]
+     */
+    let (b0, b1) = (&bits[0], &bits[1]);
+    let (a, b, c, d) = (&element, &elements[0], &elements[1], &elements[2]);
+
+    /// Define constrain macro to allow legible definition of positional constraints.
+    /// See example expansions in comment to first usages below.
+    macro_rules! witness {
+        ( $var:ident <== if $cond:ident { $a:expr } else { $b:expr }) => {
+            let $var = pick(cs.namespace(|| stringify!($var)), $cond, $a, $b)?;
+        };
+
+        // Handle conditions which are references.
+        ( $var:ident <== if &$cond:ident { $a:expr } else { $b:expr }) => {
+            let $var = pick(cs.namespace(|| stringify!($var)), &$cond, $a, $b)?;
+        };
+    }
+
+    // Declaration:
+    witness!(p0_x0 <== if b0 { b } else { a });
```

Comment:
Maybe change "p" to "pick" in the variable names and clarify what "x0" means?

Reply:
'p' is for position, actually. I added an example reading of the witness naming to set the reader on the right path.